hazelcast/src/main/java/com/hazelcast/executor/impl/ExecutorServiceProxy.java

```diff
@@ -372,12 +376,12 @@ private <T> void submitToPartitionOwner(@Nonnull Callable<T> task,
                 .createInvocationBuilder(DistributedExecutorService.SERVICE_NAME, op, partitionId)
                 .invoke();
         if (callback != null) {
-            future.whenCompleteAsync(new ExecutionCallbackAdapter<>(callback))
+            future.whenCompleteAsync(new ExecutionCallbackAdapter<>(callback), internalAsyncExecutor)
```

Comment:
Thanks, Vassilis for the valuable point!
I think it could be the `CALLER_RUNS`, so the user code should be executed in a user thread.
Or we can create one more executor for the user code.

Reply:
Agree with Vassilis, user facing api can continue to use common FJP.